.github/workflows/tests.yml

```diff
@@ -0,0 +1,93 @@
+name: Tests
+on:
+  pull_request:
+  push:
+    branches:
+    - master
+  schedule:
+    # Weekly.
+    - cron: "0 0 * * 0"
+
+jobs:
+  track-config:
+    name: Check track configuration
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Fetch
+        run: bin/fetch-configlet
+      - name: Lint
+        run: bin/configlet lint .
+      - name: Ensure READMEs are updated
+        run: sh ./bin/ensure-readmes-are-updated.sh
+      - name: Check configlet fmt
+        run: sh ./bin/check-configlet-fmt.sh
+      - name: Install yq (for stack resolvers)
+        run: |
+         sudo add-apt-repository -y ppa:rmescandon/yq
+         sudo apt-get -q update
+         sudo apt-get -y install yq
+      - name: Ensure stack resolvers synced
+        run: sh ./bin/ensure-stack-resolvers-synced.sh
+      - name: Check for invalid UUIDs
+        # can be removed once `configlet lint` gains this ability.
+        # Check issue https://github.com/exercism/configlet/issues/99
+        run: |
+          uuids=$(jq --raw-output '.exercises | map(.uuid) | .[]' config.json)
+          bad_uuid=$(echo "$uuids" | grep -vE '^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$' || true)
+          if [ -n "$bad_uuid" ]; then
+            echo "invalid UUIDs found! please correct these to be valid UUIDs:"
+            echo "$bad_uuid"
+            exit 1
+          fi
+  exercises:
+    name: Check exercises
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        resolver:
+          - lts-from-exercises
+          - nightly
+    # Discussion seems to indicate that continue-on-error is not exactly equivalent to Travis's allow_failures, but it's the best we can do for now.
+    # If failures on nightly cause too much confusion in CI status,
+    # consider just not running against the nightly resolver.
+    # https://github.com/actions/toolkit/issues/399
+    continue-on-error: ${{ matrix.resolver == 'nightly' }}
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/cache@v2
+        with:
+          # A little unfortunate:
+          # A change in any one exercise's package.yaml invalidates the cache for all other exercises.
+          # We are not sure there is a way around that or a better way to do things.
+          # Reason:
+          # By the https://github.com/actions/cache/blob/main/src/save.ts,
+          # the cache is not saved if it has already been generated.
+          # Thus the cache would slowly drift out of date unless we key on package.yaml contents.
+          key: ${{ matrix.resolver }}-${{ hashFiles('exercises/*/package.yaml') }}
+          path: |
+            ~/.stack
+            ~/.foldercache
```

Comment:
Can you verify / correct my assumptions:
 - Travis currently caches all external dependencies but completely rebuilds all exercise projects,
 - The bin/test-stub script creates a build dir shared between an exercise's project and its examples,
 - We can't easily achieve this with actions/cache@v2 because the cache can't be invalidated per exercise?

What if:
 - We add the current week number ([somewhat close example](https://github.com/actions/cache#creating-a-cache-key)) to the cache key,
 - We add .stack-work/ for each exercise as part of the cache path?
   (That'd be in bin/test-stub's `$buildfolder`, I suppose?)

Then CI will only rebuild affected modules in any exercise, and we won't prove that it can be built from scratch. But any change to package.yaml, the stub, an example solution, stack.yaml, or the like would trigger a rebuild of the affected parts.

And weekly we'd get a complete rebuild in case something slipped `stack`.

Alternatively, what if we add .stack-work/ for each exercise as part of the cache part, but we run a `stack clean` before `stack build` / `stack test`?

Reply:
(Side note: Doesn't `hashFiles('exercises/*/package.yaml')` refer only to exercise project files and not the ones located underneath examples/? So in case an example solution is changed, this wouldn't trigger a cache rebuild in the same way that a change in the exercise's package.yaml would.)

Reply:
> Doesn't `hashFiles('exercises/*/package.yaml')` refer only to exercise project files and not the ones located underneath examples/

Hmm, you are right. It looks like I should use `exercises/**/package.yaml` instead, then.

Reply:
> Travis currently caches all external dependencies but completely rebuilds all exercise projects,

I think that is true, depending on how complete we're talking. I'll explain more below.

> The bin/test-stub script creates a build dir shared between an exercise's project and its examples,

I think we managed to keep all of the build dirs separate, but it's been a while. I am seeing we have:
https://github.com/exercism/haskell/blob/3b9167ce1686821b65c6a99502d1622bd249a83f/bin/test-stub#L41-L43
https://github.com/exercism/haskell/blob/3b9167ce1686821b65c6a99502d1622bd249a83f/bin/test-example#L42-L44
(examplename is "stub" for the stub).
So the intent of all that is that all the examples+stub get their own separate .stack-work, but that because of the symlink they all end up in ~/.foldercache, where they'll all get cached.

> We can't easily achieve this with actions/cache@v2 because the cache can't be invalidated per exercise?

Well, this is an interesting question. For each `step` that's an instance of `actions/cache@v2`, I get to provide a list of paths but only one cache key. The one cache key is why I cannot say the cache is only valid for this exercise but still valid for these other ones. I have not experimented with whether it works to provide multiple instances of `action/cache@v2` in the same job, but I think that has its own questions:

* If we want one instance of `action/cache` per exercise, that may imply editing this workflow whenever an exercise is added? Or some other idea I haven't thought of.
* Say two exercises depend on the same external dependency - ideally it'd remain possible to only build that dependency once between those two exercises. Not sure whether that remains possible (could be, just haven't thought about it).

> What if:
> We add the current week number to the cache key,

Good idea. I think periodically having a completely clean rebuild is probably helpful. And the fact that a weekly build happens at the start of a week means that most of the time, we should get cache hits and not notice the complete rebuilds.

> We add .stack-work/ for each exercise as part of the cache path?

I think that got achieved by the symlinking above.